wire-ios/Wire-iOS/Sources/Analytics/TrackingManager/TrackingManager.swift

```diff
@@ -48,21 +48,36 @@ final class TrackingManager: NSObject, TrackingInterface {
             ExtensionSettings.shared.disableAnalyticsSharing
         }
         set {
-            ExtensionSettings.shared.disableAnalyticsSharing = newValue
-            AVSFlowManager.getInstance()?.setEnableMetrics(!newValue)
-
-            do {
-                if newValue {
-                    let disableUseCase = try sessionManager.makeDisableAnalyticsSharingUseCase()
-                    disableUseCase.invoke()
+            self.showAnalyticsConsentAlert { userConsented in
+                if userConsented {
+                    self.updateAnalyticsSharing(!newValue)
                 } else {
-                    let enableUseCase = try sessionManager.makeEnableAnalyticsSharingUseCase()
-                    enableUseCase.invoke()
+                    // User rejected, so we keep analytics disabled
+                    self.updateAnalyticsSharing(true)
                 }
-            } catch {
-                print("Failed to toggle analytics sharing: \(error)")
             }
         }
     }
 
+    private func updateAnalyticsSharing(_ disable: Bool) {
+        ExtensionSettings.shared.disableAnalyticsSharing = disable
+        AVSFlowManager.getInstance()?.setEnableMetrics(!disable)
+
+        do {
+            if disable {
+                let disableUseCase = try sessionManager.makeDisableAnalyticsUseCase()
+                disableUseCase.invoke()
+            } else {
+                let enableUseCase = try sessionManager.makeEnableAnalyticsUseCase()
+                enableUseCase.invoke()
+            }
+        } catch {
+            WireLogger.analytics.error("Failed to toggle analytics sharing: \(error)")
+
+            // Reset the setting
+            ExtensionSettings.shared.disableAnalyticsSharing = !disable
+            AVSFlowManager.getInstance()?.setEnableMetrics(disable)
```

Comment:
This is error prone because you're assuming that the value was previously the opposite of the new value but there's no guard that is actually the case. 

Instead, you could invoke the use case and only after a success update the these settings.

Reply:
Done: https://github.com/wireapp/wire-ios/pull/1767/commits/4d057e135cfeef971c25594ac2e148f186b7e24f